File: apps/go/routes/body/blob.go
//go:build js && wasm

package main

import "github.com/primate-run/go/route"

var _ = route.With{ContentType: route.Blob}.Post(func(request route.Request) any {
	blob, err := request.Body.Blob()
	if err != nil {
		return map[string]any{"error": err.Error()}
	}

	head := []int{}
	for i := 0; i < len(blob.Data) && i < 4; i++ {
		head = append(head, int(blob.Data[i]))
	}

	return map[string]any{
		"type": blob.Type,
		"size": len(blob.Data),
		"head": head,
	}
})

File: apps/js/routes/body/blob.js
import route from "primate/route";

export default route({
  async post(request) {
    const blob = await request.body.blob();
    const buf = new Uint8Array(await blob.arrayBuffer());

    return {
      head: Array.from(buf.slice(0, 4)),
      size: buf.byteLength,
      type: blob.type || (request.headers.try("content-type") ??
        "application/octet-stream"),
    };
  },
});
